File: pyglet_client/test/camera.py
from math import sin, cos, pi

from pyglet.gl import *
from pyglet import clock, font, image, window

class Camera(object):

    def __init__(self, win, x=0.0, y=0.0, z=0.0, rot=0.0, zoom=1.0):
        self.win = win
        self.x = x
        self.y = y
        self.z = z
        self.x_angle = 0.0
        self.y_angle = 0.0
        self.rot = rot
        self.zoom = zoom
        self.rts = True

    def worldProjection(self):
        glMatrixMode(GL_PROJECTION)
        glLoadIdentity()
        aspect = self.win.width / float(self.win.height)
        #gluPerspective( 45.0 / self.scale, aspect, 0.1, 100.0);
        gluPerspective(65, aspect, .1, 1000)


        glMatrixMode( GL_MODELVIEW )
        glLoadIdentity()

        #camera_focus_x = self.x + cos( self.x_angle * pi)
        #camera_focus_y = self.y + sin( self.x_angle * pi)
        #camera_focus_z = self.z + sin( self.y_angle)

        camera_focus_x = self.x + cos( self.x_angle * pi) * cos( self.y_angle * pi)
        camera_focus_y = self.y + sin( self.x_angle * pi) * cos( self.y_angle * pi)
        camera_focus_z = self.z + sin( self.y_angle)

        gluLookAt( self.x, self.y, self.z,
                camera_focus_x, camera_focus_y, camera_focus_z,
                0., 0., 1.0)

        glEnable (GL_DEPTH_TEST)
        #glEnable(GL_CULL_FACE);

    def hudProjection(self):
        glMatrixMode(GL_PROJECTION)
        glLoadIdentity()
        gluOrtho2D(0, self.win.width, 0, self.win.height)

        glMatrixMode( GL_MODELVIEW )
        glLoadIdentity()

        glDisable(GL_DEPTH_TEST);
        #glDisable(GL_CULL_FACE);


    def move_camera(self, dx, dy, dz):

        if self.rts == True:
            #dx delta
            self.x += dx*cos(self.x_angle * pi)
            self.y += dx*sin(self.x_angle * pi)
            #dy delta
            self.x += dy*cos(self.x_angle * pi + pi/2)
            self.y += dy*sin(self.x_angle * pi + pi/2)
        else:
            self.x += dx
            self.y += dy
        #dz delta
        self.z += dz

    def pan(self, dx_angle, dy_angle):
        #print "dx_angle= " + str(dx_angle)
        #print "dy_angle= " + str(dy_angle)
        self.x_angle += dx_angle
        self.y_angle += dy_angle
        if self.y_angle < -0.499:
            self.y_angle = -0.499
        if self.y_angle > 0.499:
            self.y_angle = 0.499

    def on_resize(self, width, height):
        print "Resize Window"
        glViewport(0, 0, width, height)
        self.worldProjection()
        return pyglet.event.EVENT_HANDLED

class Hud(object):

    def __init__(self, win):
        self.win = win
        helv = font.load('Helvetica', win.width / 15.0)
        self.text = font.Text(
            helv,
            'Hello, World!',
            x=win.width / 2,
            y=win.height / 2,
            halign=font.Text.CENTER,
            valign=font.Text.CENTER,
            color=(1, 1, 1, 0.5),
        )
        self.fps = clock.ClockDisplay()
        self._init_reticle()

    def _init_reticle(self):
        self.reticle = pyglet.image.load('./target.png')
        self.reticle_texture = self.reticle.get_texture()

        rh = 16.
        rw = 16.

        w = float(self.win.width)/2
        h = float(self.win.height)/2
        print str((h,w))
        x_min = w - rw/2
        x_max = w + rw/2
        y_min = h - rh/2
        y_max = h + rh/2

        v_list = [
        x_min, y_max,
        x_max, y_max,
        x_max, y_min,
        x_min, y_min
        ]
        print str(v_list)

        self.reticleVertexList = pyglet.graphics.vertex_list(4,
            ('v2f\static', v_list),
            ('c3B\static', (255,255,255) *4),
            ("t3f\static", self.reticle_texture.tex_coords),
            )

    def draw(self):
        #self.text.draw()
        self.draw_reticle()
        self.fps.draw()

    def draw_reticle(self):
        glEnable(GL_TEXTURE_2D)        # typically target is GL_TEXTURE_2D
        glBindTexture(self.reticle_texture.target, self.reticle_texture.id)

        glEnable (GL_BLEND)
        glBlendFunc (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
        self.reticleVertexList.draw(pyglet.gl.GL_QUADS)
        glDisable (GL_BLEND);

File: pyglet_client/netclient/SDL/camera.c
/* moved to header
#include <SDL.h>
#include <SDL_opengl.h>

#include <math.h>

struct Camera {
float fov;
float x_size;
float y_size;
float z_near;
float z_far;

float x,y,z;
float x_angle;
float y_angle;
};
*/

#include "./camera.h"

#define PI 3.14159265

/*
void _camera_projection( Camera c) {
    float aspect = c.x_size / c.y_size;
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective( c.fov, aspect, c.z_near, c.z_far);
}
*/

struct Camera* camera;

float * model_view_matrix;

void set_model_view_matrix(float *a){
    model_view_matrix = a;
}

int _world_projection(struct Camera* c) {
    float aspect = c->x_size / c->y_size;
    float camera_focus_x,  camera_focus_y,  camera_focus_z;
    float length;
    c->ratio = c->x_size / c->y_size;

    glClear( GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT );

    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluPerspective( c->fov, aspect, c->z_near, c->z_far);

    glMatrixMode( GL_MODELVIEW );
    glLoadIdentity();

    c->xl = c->x + cos( c->x_angle * PI) * cos( c->y_angle * PI);
    c->yl = c->y + sin( c->x_angle * PI) * cos( c->y_angle * PI);
    c->zl = c->z + sin( c->y_angle);

    length = sqrt(c->xl*c->xl + c->yl*c->yl + c->zl*c->zl);
    //c->xl /= length;
    //c->yl /= length;
    //c->zl /= length;

    c->xu = 0;
    c->yu = 0;
    c->zu = 1;

    gluLookAt(c->x,c->y,c->z,
               c->xl, c->yl, c->zl,
               c->xu, c->yu, c->zu);

    if(model_view_matrix != NULL){
        glGetFloatv(GL_MODELVIEW_MATRIX, model_view_matrix);
    } else {
        printf("!? null pointer \n");
    }

    //printf( "(%f, %f, %f), (%f, %f, %f) \n",c.x,c.y,c.z,camera_focus_x, camera_focus_y,  camera_focus_z );

    //glEnable(GL_TEXTURE_2D);

    glEnable (GL_DEPTH_TEST);
    //glEnable(GL_CULL_FACE);

/*
        glEnable (GL_DEPTH_TEST)
        #glEnable(GL_CULL_FACE);
*/
    //printf( "gl error: %s\n", SDL_GetError() );
    return 0;
}

void __inline end_world_projection() {
   glDisable (GL_DEPTH_TEST);
}

int _hud_projection(struct Camera * c) {

end_world_projection();

glMatrixMode(GL_PROJECTION);
glLoadIdentity();
gluOrtho2D(0, c->x_size, 0, c->y_size);

//printf( "(%f, %f) \n", c.x_size, c.y_size);

glMatrixMode( GL_MODELVIEW );
glLoadIdentity();

glDisable(GL_DEPTH_TEST);

glEnable(GL_TEXTURE_2D);
//glEnable(GL_TEXTURE_2D);

/*
         glMatrixMode(GL_PROJECTION)
        glLoadIdentity()
        gluOrtho2D(0, self.win.width, 0, self.win.height)

        glMatrixMode( GL_MODELVIEW )
        glLoadIdentity()

        glDisable(GL_DEPTH_TEST);
        #glDisable(GL_CULL_FACE);
        glEnable(gl.GL_TEXTURE_2D)
*/
return 0;
}

struct Camera* _get_camera() {
    return camera;
}
